pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,435 @@
+					}
+				}
+
+				err := s.pushDeltaVirtualHost(con, s.globalPushContext(), unsubscribeHosts)
+				if err != nil {
+					return err
+				}
+
+			case EndpointType:
+				return status.Errorf(codes.Unimplemented, "not implemented")
+			default:
+				adsLog.Warnf("ADS: Unknown watched resources %s", discReq.String())
+			}
+
+			if !con.added {
+				con.added = true
+				s.addCon(con.ConID, con)
+				defer s.removeCon(con.ConID, con)
+			}
+		case pushEv := <-con.pushChannel:
+			// It is called when config changes.
+			// This is not optimized yet - we should detect what changed based on event and only
+			// push resources that need to be pushed.
+
+			// TODO: possible race condition: if a config change happens while the envoy
+			// was getting the initial config, between LDS and RDS, the push will miss the
+			// monitored 'routes'. Same for CDS/EDS interval.
+			// It is very tricky to handle due to the protocol - but the periodic push recovers
+			// from it.
+
+			err := s.pushConnection(con, pushEv)
+			if err != nil {
+				return nil
+			}
+
+		}
+	}
+}
+
+/*func mergeSubscribeUnsubsribe(resource []string, subscribe []string, unsubscribe []string) []string {
+	for i := 0; i < len(resource); i++ {
+		for j := 0; j < len(subscribe); j++ {
+			for k := 0; k < len(unsubscribe); k++ {
+
+			}
+		}
+	}
+}*/
+
+// Send with timeout
+func (conn *XdsConnection) sendDelta(res *xdsapi.DeltaDiscoveryResponse) error {
+	done := make(chan error)
+	t := time.NewTimer(SendTimeout)
+	go func() {
+		err := conn.deltaStream.Send(res)
+		done <- err
+		conn.mu.Lock()
+		if res.Nonce != "" {
+			conn.RouteNonceSent = res.Nonce
+		}
+		conn.mu.Unlock()
+	}()
+	select {
+	case <-t.C:
+		// TODO: wait for ACK
+		adsLog.Infof("Timeout writing %s", conn.ConID)
+		xdsResponseWriteTimeouts.Add(1)
+		return errors.New("timeout sending")
+	case err := <-done:
+		t.Stop()
+		return err
+	}
+}
+
+func subscribeToVHost(sidecarScope *model.SidecarScope, pushContext *model.PushContext, routeConfigName uint32, host string, namespace string) (*model.SidecarScope, error) {
+	if sidecarScope == nil {
+		return nil, fmt.Errorf("nil sidecar scope")
+	}
+
+	var sidecarConfig *networking.Sidecar
+	var ok bool
+
+	if sidecarScope.Config != nil {
+		sidecarCfg := sidecarScope.Config
+		sidecarConfig, ok = sidecarCfg.Spec.(*networking.Sidecar)
+		if !ok {
+			return nil, fmt.Errorf("invalid sidecar config")
+		}
+	} else {
+		sidecarConfig = &networking.Sidecar{
+			Egress: []*networking.IstioEgressListener{
+				{
+					//Bavery_QUESTION: This is likely wrong, but how do we figure out the port info? Specifically the protocol
+					Port: &networking.Port{
+						Number:   routeConfigName,
+						Protocol: "HTTP",
+					},
+				},
+			},
+		}
+
+	}
+
+	found := false
+
+	for _, egressPb := range sidecarConfig.Egress {
+		if egressPb.Port.Number != uint32(routeConfigName) {
+			continue
+		}
+		found = true
+		egressPb.Hosts = append(egressPb.Hosts, host)
```

Comment:
??? why are you adding data plane supplied hosts to user generated config? there is no source of truth anymore

Reply:
+1

The dynamic hosts should be in a separate list I think